plugins/net_plugin/net_plugin.cpp

```diff
@@ -2661,56 +2727,33 @@ namespace eosio {
             fc_dlog( logger, "Skipping connect due to go_away reason ${r}",("r", reason_str( no_retry )));
             return false;
       }
-
-      string::size_type colon = peer_address().find(':');
-      if (colon == std::string::npos || colon == 0) {
-         fc_elog( logger, "Invalid peer address. must be \"host:port[:<blk>|<trx>]\": ${p}", ("p", peer_address()) );
-         return false;
-      }
-
-      connection_ptr c = shared_from_this();
-
       if( consecutive_immediate_connection_close > def_max_consecutive_immediate_connection_close || no_retry == benign_other ) {
          fc::microseconds connector_period = my_impl->connections.get_connector_period();
          fc::lock_guard g( conn_mtx );
          if( last_close == fc::time_point() || last_close > fc::time_point::now() - connector_period ) {
-            return true; // true so doesn't remove from valid connections
+            return true;
          }
       }
-
+      connection_ptr c = shared_from_this();
       strand.post([c]() {
-         auto [host, port, type] = split_host_port_type(c->peer_address());
-         c->set_connection_type( c->peer_address() );
-
-         auto resolver = std::make_shared<tcp::resolver>( my_impl->thread_pool.get_executor() );
-         connection_wptr weak_conn = c;
-         resolver->async_resolve(host, port, boost::asio::bind_executor( c->strand,
-            [resolver, weak_conn, host = host, port = port]( const boost::system::error_code& err, const tcp::resolver::results_type& endpoints ) {
-               auto c = weak_conn.lock();
-               if( !c ) return;
-               if( !err ) {
-                  c->connect( resolver, endpoints );
-               } else {
-                  fc_elog( logger, "Unable to resolve ${host}:${port} ${error}",
-                           ("host", host)("port", port)( "error", err.message() ) );
-                  c->set_state(connection_state::closed);
-                  ++c->consecutive_immediate_connection_close;
-               }
-         } ) );
-      } );
+         auto it = my_impl->connections.get_connection_iterator(c);
+         c->connect(it->ips, it);
+      });
       return true;
    }
 
    // called from connection strand
-   void connection::connect( const std::shared_ptr<tcp::resolver>& resolver, const tcp::resolver::results_type& endpoints ) {
+   void connection::connect( const tcp::resolver::results_type& endpoints,
+                             connections_manager::connection_details_index::const_iterator conn_details ) {
```

Comment:
`conn_details` is an iterator into a collection that requires a lock since it can be accessed from multiple threads. This iterator is being passed off to a connection strand to be accessed via the lambda. That lambda is called without the lock. Even if you acquired the lock the iterator is not necessarily any good. Once the lock is released the iterator should be considered invalid.

Best you could do is re-find the connection via a find with a locked mutex.

Reply:
Reworked how `connection::reconnect()` works to delegate back to `connections_manager`, eliminating this break in encapsulation entirely.